File: backend/src/ai/utils/logger.py
import os
import json
import traceback
from datetime import datetime

# Resolve logs directory to d:\my-projects\kairos\backend\logs
LOG_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../logs"))

if not os.path.exists(LOG_DIR):
    os.makedirs(LOG_DIR, exist_ok=True)

SEPARATOR = "\n------------------------------------------------------------\n"

def write_log(category: str, content):
    try:
        timestamp = datetime.now().isoformat()
        log_file_path = os.path.join(LOG_DIR, f"{category}.log")
        
        log_string = f"[{timestamp}]\n"
        if isinstance(content, (dict, list)):
            log_string += json.dumps(content, indent=2, default=str)
        elif isinstance(content, BaseException):
            log_string += "".join(traceback.format_exception(None, content, content.__traceback__))
        else:
            log_string += str(content)
            
        log_string += SEPARATOR
        
        with open(log_file_path, "a", encoding="utf-8") as f:
            f.write(log_string)
    except Exception as e:
        print(f"Failed to write to log {category}: {e}")

File: backend/src/utils/logger.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Resolve logs directory at the root of the backend folder
const LOG_DIR = path.resolve(__dirname, '../../../logs');

// Ensure log directory exists
if (!fs.existsSync(LOG_DIR)) {
    fs.mkdirSync(LOG_DIR, { recursive: true });
}

const SEPARATOR = '\n------------------------------------------------------------\n';

/**
 * Write a log to a specific category file with a separator snippet.
 * @param {string} category - e.g., 'ai', 'prompting', 'overpass', 'services', 'geo_api'
 * @param {string|object} content - The content to log
 */
export function writeLog(category, content) {
    try {
        const timestamp = new Date().toISOString();
        const logFilePath = path.join(LOG_DIR, `${category}.log`);

        let logString = `[${timestamp}]\n`;
        if (typeof content === 'object') {
            // Handle Error objects properly
            if (content instanceof Error) {
                logString += `${content.name}: ${content.message}\n${content.stack}`;
            } else {
                logString += JSON.stringify(content, null, 2);
            }
        } else {
            logString += content;
        }
        logString += SEPARATOR;

        fs.appendFileSync(logFilePath, logString, 'utf8');
    } catch (error) {
        console.error(`Failed to write to log ${category}:`, error);
    }
}

export default { writeLog };
